### PR TITLE
Issue27 visible sections

### PR DESCRIPTION
**Pull request by Malarg**

### CODE REVIEW

File: lib/src/code/code.dart [hunk 1]
```diff
@@ -92,15 +95,32 @@ class Code {
...
+
+    final visibleSectionsHiddenRanges = _visibleSectionsToHiddenRanges(
+      visibleSectionsNames,
+      sectionsMap,
+      lines,
+    );
+
+    final commentsHiddenRanges = _commentsToHiddenRanges(
+      serviceComments,
+      visibleSectionsHiddenRanges,
```
Comment (alexeyinkin) [inline lib/src/code/code.dart:117]: Revert, the builder can merge overlapping ranges.

File: lib/src/code/code.dart [hunk 2]
```diff
@@ -109,6 +129,7 @@ class Code {
     final hiddenLineRangesBuilder = HiddenLineRangesBuilder(
       codeLines: lines,
       hiddenRanges: hiddenRanges,
+      shouldReduceStartToOne: isCodeReadonly,
```
Comment (alexeyinkin) [inline lib/src/code/code.dart:132]: Remove.

File: lib/src/code/code.dart [hunk 3]
```diff
@@ -177,19 +209,96 @@ class Code {
     }
   }
 
+  ///Assumes that there are only one visible section.
```
Comment (alexeyinkin) [inline lib/src/code/code.dart:212]: there is

File: lib/src/code/code.dart [hunk 4]
```diff
@@ -177,19 +209,96 @@ class Code {
...
+        .firstOrNull;
+
+    final result = <String, HiddenRange>{};
+
+    if (section == null) {
+      return result;
+    }
+
+    late int beginnigEnd;
+    late int beginnigLastLine;
```
Comment (alexeyinkin) [inline lib/src/code/code.dart:232]: Remove 'late'

File: lib/src/code/code.dart [hunk 5]
```diff
@@ -177,19 +209,96 @@ class Code {
...
+    final startLine = lines.lines[section.startLine];
+    if (startLine.text.substring(startLine.indent).startsWith('// [START')) {
```
Comment (alexeyinkin) [inline lib/src/code/code.dart:234]: Remove the two branches. The visible section always starts on the line where the section starts.

File: lib/src/code/code.dart [hunk 6]
```diff
@@ -177,19 +209,96 @@ class Code {
...
+        lines.lines.last.textRange.end,
+        firstLine: section.endLine!,
+        lastLine: section.endLine!,
+        wholeFirstLine: false,
+      );
+    }
+    return result;
+  }
+
   static Map<int, HiddenRange> _commentsToHiddenRanges(
```
Comment (alexeyinkin) [inline lib/src/code/code.dart:262]: Revert the whole method change?

File: lib/src/hidden_ranges/hidden_line_ranges.dart [hunk 7]
```diff
@@ -6,11 +6,13 @@ class HiddenLineRanges with EquatableMixin {
   final List<LineNumberingBreakpoint> breakpoints;
```
Comment (alexeyinkin) [inline lib/src/hidden_ranges/hidden_line_ranges.dart:6]: Revert the whole file?

File: lib/src/hidden_ranges/visible_section.dart [hunk 8]
```diff
@@ -0,0 +1,47 @@
+import '../../flutter_code_editor.dart';
```
Comment (alexeyinkin) [inline lib/src/hidden_ranges/visible_section.dart:1]: Delete file.

File: lib/src/single_line_comments/single_line_comment.dart [hunk 9]
```diff
@@ -23,6 +23,8 @@ class SingleLineComment {
 
   late final isReadonly = _checkIfReadonly();
 
+  int get lastCharacterIndex => characterIndex + outerContent.length;
```
Comment (alexeyinkin) [inline lib/src/single_line_comments/single_line_comment.dart:26]: Remove?

Comment (alexeyinkin) [reply lib/src/single_line_comments/single_line_comment.dart:26]: If we keep it, it is `endCharacterIndex`

File: test/src/visible_sections/visible_sections_test.dart [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+      expect(controller.value.text, _method2n3);
+      expect(
+        controller.code.hiddenLineRanges.visibleLineNumbers.toList(),
+        [0, 1, 2, 3, 4, 5, 6, 7],
+      );
+    });
+  });
+}
+
+CodeController _buildController({
```
Comment (alexeyinkin) [inline test/src/visible_sections/visible_sections_test.dart:194]: Maybe reuse `createController` from `create_app.dart`?

**Review by alexeyinkin — CHANGES_REQUESTED** [review]
Thanks.

File: lib/src/code/code.dart (hunk 1 above)
Comment (Malarg) [reply lib/src/code/code.dart]: +

File: lib/src/code/code.dart (hunk 2 above)
Comment (Malarg) [reply lib/src/code/code.dart]: +

File: lib/src/code/code.dart [hunk 11]
```diff
@@ -155,9 +178,18 @@ class Code {
...
-  static void _applyNamedSectionsToLines({
+  static void _makeCodeReadonly({
+    required List<CodeLine> lines,
+  }) {
+    for (int i = 0; i < lines.length; i++) {
+      lines[i] = lines[i].copyWith(isReadOnly: true);
+    }
+  }
+
+  static void _applyReadonlySectionsToLines({
```
[the remark this answers is not in the document]
Comment (Malarg) [reply lib/src/code/code.dart]: +

File: lib/src/code/code.dart (hunk 3 above)
Comment (Malarg) [reply lib/src/code/code.dart]: +

File: lib/src/code/code.dart (hunk 4 above)
Comment (Malarg) [reply lib/src/code/code.dart]: +

File: lib/src/code/code.dart (hunk 5 above)
Comment (Malarg) [reply lib/src/code/code.dart]: +

File: lib/src/hidden_ranges/hidden_line_ranges.dart (hunk 7 above)
Comment (Malarg) [reply lib/src/hidden_ranges/hidden_line_ranges.dart]: +

File: lib/src/hidden_ranges/hidden_range.dart [hunk 12]
```diff
@@ -1,6 +1,7 @@
 import 'package:equatable/equatable.dart';
```
[the remark this answers is not in the document]
Comment (Malarg) [reply lib/src/hidden_ranges/hidden_range.dart]: +

File: lib/src/hidden_ranges/visible_section.dart (hunk 8 above)
Comment (Malarg) [reply lib/src/hidden_ranges/visible_section.dart]: +

File: lib/src/single_line_comments/single_line_comment.dart (hunk 9 above)
Comment (Malarg) [reply lib/src/single_line_comments/single_line_comment.dart]: removed

File: test/src/visible_sections/visible_sections_test.dart (hunk 10 above)
Comment (Malarg) [reply test/src/visible_sections/visible_sections_test.dart]: +

**Review by alexeyinkin — CHANGES_REQUESTED** [review]

File: lib/src/code/code.dart [hunk 13]
```diff
@@ -177,6 +207,44 @@ class Code {
...
+        lastLine: section.firstLine,
+        wholeFirstLine: true,
+      );
+    }
+    if (section.lastLine != null && section.lastLine! < lines.lines.length - 1) {
+      result['ending'] = HiddenRange(
+        lines.lines[section.lastLine!].textRange.end,
+        lines.lines.last.textRange.end,
+        firstLine: section.lastLine!,
+        lastLine: section.lastLine!,
```
[the remark this answers is not in the document]
Comment (alexeyinkin) [reply lib/src/code/code.dart:241]: `lastLine: lines.length - 1`?

File: lib/src/code/code.dart [hunk 14]
```diff
@@ -177,6 +207,44 @@ class Code {
...
+      return result;
+    }
+
+    final startLine = lines.lines[section.startLine];
+    if (startLine.textRange.start > 0) {
+      result['beginning'] = HiddenRange(
+        0,
+        startLine.textRange.start,
+        firstLine: 0,
+        lastLine: section.startLine,
```
[the remark this answers is not in the document]
Comment (Malarg) [reply lib/src/code/code.dart]: merged from main. now its first line

[the remark this answers is not in the document]
Comment (alexeyinkin) [reply lib/src/code/code.dart]: I mean ` - 1`

File: test/src/visible_sections/visible_sections_test.dart [hunk 15]
```diff
@@ -0,0 +1,229 @@
...
+  }
+  
+  void method() {
+  }
+}''');
+      expect(
+        controller.code.hiddenLineRanges.visibleLineNumbers.toList(),
+        [0, 1, 2, 3, 4, 5, 6, 7],
+      );
+    });
```
[the remark this answers is not in the document]
Comment (Malarg) [reply test/src/visible_sections/visible_sections_test.dart:106]: I'm not sure that its really required here. In this case code don't become more readable, but requires additional code lines

File: test/src/visible_sections/visible_sections_test.dart [hunk 16]
```diff
@@ -0,0 +1,229 @@
+import 'package:flutter_test/flutter_test.dart';
```
[the remark this answers is not in the document]
Comment (Malarg) [reply test/src/visible_sections/visible_sections_test.dart:1]: Why? Where aren't `code_controller` folder at `test/src`. But there are such folders as `named_sections` and `hidden_ranges`, which places at the top of the structure

File: lib/src/code/code.dart [hunk 17]
```diff
@@ -177,6 +207,44 @@ class Code {
...
+        wholeFirstLine: true,
+      );
+    }
+    if (section.lastLine != null && section.lastLine! < lines.lines.length - 1) {
+      result['ending'] = HiddenRange(
+        lines.lines[section.lastLine!].textRange.end,
+        lines.lines.last.textRange.end,
+        firstLine: section.lastLine!,
+        lastLine: section.lastLine!,
+        wholeFirstLine: false,
```
[the remark this answers is not in the document]
Comment (alexeyinkin) [reply lib/src/code/code.dart]: Actually it's always true.

File: lib/src/code_field/code_controller.dart [hunk 18]
```diff
@@ -522,6 +515,46 @@ class CodeController extends TextEditingController {
     );
   }
 
+  void foldCommentAtLineZero() {
```
[the remark this answers is not in the document]
Comment (Malarg) [reply lib/src/code_field/code_controller.dart]: fixed

File: lib/src/code_field/code_controller.dart [hunk 19]
```diff
@@ -512,8 +502,11 @@ class CodeController extends TextEditingController {
     _visibleSectionNames = sectionNames;
     final oldCode = _code;
     _updateCode(_code.text);
-    code.foldedAs(oldCode);
-    
+
+    _updateValueWithSelection(oldCode);
+  }
+
+  void _updateValueWithSelection(Code oldCode) {
```
[the remark this answers is not in the document]
Comment (Malarg) [reply lib/src/code_field/code_controller.dart]: fixed

**Review by alexeyinkin — APPROVED** [review]

**Pull request merged**